numpyro/infer/elbo.py

```diff
@@ -280,3 +281,266 @@ def single_particle_elbo(rng_key):
+            downstream_costs[site].add((stacks[child], model_trace[child]["log_prob"]))
+            downstream_guide_cost_nodes[site].update([child])
+
+    for k in non_reparam_nodes:
+        downstream_costs[k] = downstream_costs[k].sum_to(
+            guide_trace[k]["cond_indep_stack"]
+        )
+
+    return downstream_costs, downstream_guide_cost_nodes
+
+
+class TraceGraph_ELBO:
+    """
+    A TraceGraph implementation of ELBO-based SVI. The gradient estimator
+    is constructed along the lines of reference [1] specialized to the case
+    of the ELBO. It supports arbitrary dependency structure for the model
+    and guide as well as baselines for non-reparameterizable random variables.
+    Where possible, conditional dependency information as recorded in the
+    trace is used to reduce the variance of the gradient estimator.
+    In particular two kinds of conditional dependency information are
+    used to reduce variance:
+
+    - the sequential order of samples (z is sampled after y => y does not depend on z)
+    - :class:`~numpyro.plate` generators
+
+    References
+
+    [1] `Gradient Estimation Using Stochastic Computation Graphs`,
+        John Schulman, Nicolas Heess, Theophane Weber, Pieter Abbeel
+    [2] `Neural Variational Inference and Learning in Belief Networks`
+        Andriy Mnih, Karol Gregor
+    """
+
+    def __init__(self, num_particles=1):
+        self.num_particles = num_particles
+
+    def loss(self, rng_key, param_map, model, guide, *args, **kwargs):
+        """
+        Evaluates the ELBO with an estimator that uses num_particles many samples/particles.
+
+        :param jax.random.PRNGKey rng_key: random number generator seed.
+        :param dict param_map: dictionary of current parameter values keyed by site
+            name.
+        :param model: Python callable with NumPyro primitives for the model.
+        :param guide: Python callable with NumPyro primitives for the guide.
+        :param args: arguments to the model / guide (these can possibly vary during
+            the course of fitting).
+        :param kwargs: keyword arguments to the model / guide (these can possibly vary
+            during the course of fitting).
+        :return: negative of the Evidence Lower Bound (ELBO) to be minimized.
+        """
+
+        def single_particle_elbo(rng_key):
+            model_seed, guide_seed = random.split(rng_key)
+            seeded_model = seed(model, model_seed)
+            seeded_guide = seed(guide, guide_seed)
+            model_trace, guide_trace = get_importance_trace(
+                seeded_model, seeded_guide, args, kwargs, param_map
+            )
+
+            # XXX: different from Pyro, we don't support baseline_loss here
+            non_reparam_nodes = {
+                name
+                for name, site in guide_trace.items()
+                if site["type"] == "sample"
+                and (site["is_observed"] or not site["fn"].has_rsample)
```

Comment:
shouldn't this be `not site["is_observed"] and not site["fn"].has_rsample`? that is the defn in pyro

Reply:
Good catch! Here is [pyro implementation](https://github.com/pyro-ppl/pyro/blob/0decbe54f8660cbf453ef4915495d77f0d7fb454/pyro/poutine/trace_struct.py#L293). I guess what I wanted to code is `and not (site["is_observed"] or site["fn"].has_rsample)`. Thanks for pointing this out.